src/Sets/Zonotope.jl

```diff
@@ -107,11 +107,15 @@ struct Zonotope{N<:Real, VN<:AbstractVector{N}, MN<:AbstractMatrix{N}} <: Abstra
         @assert length(center) == size(generators, 1) "the dimension of the " *
             "center ($(length(center))) and the generators " *
             "($(size(generators, 1))) need to match"
+        G = generators
         if remove_zero_generators
-            generators = delete_zero_columns!(generators)
+            nzcol = nonzero_columns(generators)
+            if length(nzcol) < size(generators, 2)
+                G = view(generators, :, nzcol)
+            end
         end
         MT = typeof(generators)
-        new{N, VN, MT}(center, generators)
+        new{N, VN, MT}(center, G)
```

Comment:
Instead you can do (I tested it):
```julia
        return _Zonotope(center, G)
    end

    # constructor with fixed types
    function _Zonotope(center::VN, generators::MN) where {N<:Real,
                                                          VN<:AbstractVector{N},
                                                          MN<:AbstractMatrix{N}}
        return new{N, VN, MN}(center, generators)
    end
```

Reply:
ok i did that change, thanks.

---

as commented above, i still consider that it's worth reviewing this default check, if we want to keep it or not. it imposes a high slowdown for the creation of zonotopes, easily 10x up to 50x from tests that i did. recall that the motivation for this constructor check (https://github.com/JuliaReach/LazySets.jl/pull/1246) was the conversion from flat hyperrectangles to zonotopes. however this can be done done like so:

```julia
function convert(::Type{Zonotope}, H::AbstractHyperrectangle)
    G = genmat(H)
    if isflat(H)
         # get nonzero columns
         # new G wih removed generators
    end
    return Zonotope(center(H), G)
end
```

Reply:
i added a proof of this principle in https://github.com/JuliaReach/LazySets.jl/pull/2072

Reply:
> it imposes a high slowdown for the creation of zonotopes

In internal code we should deactivate the check as long as we can guarantee that nothing bad happens. Your flat hyperrectangle conversion is a good example. But for the general API I would keep the checks to avoid confusion about unexpected behavior. For performance-critical code there is an option to deactivate the check.

> easily 10x up to 50x

A factor is not a fair comparison because the rest of the constructor is constant.

Reply:
> A factor is not a fair comparison because the rest of the constructor is constant.

yes, but the application that i have in mind is reachability analysis where you create thousands or hundreds of thousands of zonotopes.

Reply:
> For performance-critical code there is an option to deactivate the check.

Reply:
yes, [that(https://github.com/JuliaReach/LazySets.jl/pull/2068#discussion_r392668546 is something you proposed already and now for a 2nd time :)  i agree about the premise of deactivating this check internally so long as the operation does not create zero generators. but thinking this through, i don't know of any meaningful operation / use that would create zero generators, so i was thinking if we should rather remove this option altogether (in the constructor). the flat hyperrectangle case can be handled separately as stated above.

Reply:
> i don't know of any meaningful operation / use that would create zero generators

The user interface for creating a `Zonotope` lets you pass an arbitrary matrix.